File: src/main/kotlin/com/nikuagent/service/CliLlmClient.kt
package com.nikuagent.service

import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Claude Code CLI(`claude`)를 subprocess로 실행하는 LLM 클라이언트.
 *
 * 프롬프트는 stdin으로 전달한다 (`claude --print`는 인수 없이 실행 후 stdin에서 읽음).
 * claude CLI가 로그인된 상태여야 동작한다.
 *
 * @param binaryPath  claude 실행 파일 경로 (예: /usr/local/bin/claude)
 * @param model       사용할 모델 (null이면 CLI 기본값 사용)
 */
class CliLlmClient(
    private val binaryPath: String,
    private val model: String? = null,
) : LlmClient {

    override fun complete(prompt: String, onChunk: ((String) -> Unit)?): String {
        val cmd = buildList {
            add(binaryPath)
            add("--print")
            if (!model.isNullOrBlank()) {
                add("--model")
                add(model)
            }
            // 프롬프트는 인수가 아닌 stdin으로 전달 (긴 프롬프트 안정성 + 특수문자 이슈 방지)
        }

        val process = try {
            ProcessBuilder(cmd)
                .redirectErrorStream(true)
                .start()
        } catch (e: Exception) {
            throw LlmException(
                "Claude CLI 실행 실패: ${e.message}\n" +
                "› claude CLI가 설치되어 있는지 확인해주세요. (https://claude.ai/download)",
                e
            )
        }

        // 프롬프트를 stdin으로 write 후 즉시 close (EOF 전송)
        try {
            process.outputStream.bufferedWriter().use { it.write(prompt) }
        } catch (e: Exception) {
            process.destroyForcibly()
            throw LlmException("Claude CLI stdin 전송 실패: ${e.message}", e)
        }

        return if (onChunk != null) {
            readStreaming(process, onChunk)
        } else {
            readBlocking(process)
        }
    }

    /**
     * 스트리밍 모드: stdout을 읽으면서 onChunk 콜백 호출.
     * watchdog 스레드로 TIMEOUT_SECONDS 후 강제 종료 보장.
     */
    private fun readStreaming(process: Process, onChunk: (String) -> Unit): String {
        val sb = StringBuilder()

        // stdout 읽기 스레드
        val readerThread = Thread {
            try {
                val reader = process.inputStream.bufferedReader()
                val buffer = CharArray(512)
                var charsRead: Int
                while (reader.read(buffer).also { charsRead = it } != -1) {
                    synchronized(sb) { sb.append(buffer, 0, charsRead) }
                    onChunk(synchronized(sb) { sb.toString() })
                }
            } catch (_: Exception) { }
        }
        readerThread.isDaemon = true
        readerThread.start()

        val finished = process.waitFor(TIMEOUT_SECONDS, TimeUnit.SECONDS)
        readerThread.join(2_000)

        if (!finished) {
            process.destroyForcibly()
            val partial = synchronized(sb) { sb.toString().trim() }
            if (partial.isNotBlank()) {
                onChunk("$partial\n\n⚠️ 응답 시간 초과로 부분 결과만 표시됩니다.")
                return partial
            }
            throw LlmException(
                "Claude CLI 응답 시간 초과 (${TIMEOUT_SECONDS}초)\n" +
                "› 네트워크 상태나 claude 로그인 여부를 확인해주세요."
            )
        }

        val output = synchronized(sb) { sb.toString().trim() }
        checkNotLoggedIn(output)

        val exitCode = runCatching { process.exitValue() }.getOrDefault(-1)
        if (exitCode != 0 && output.isBlank()) {
            throw LlmException("Claude CLI 오류 (exit $exitCode). Settings에서 버전 확인 버튼으로 상태를 확인해주세요.")
        }

        return output.ifBlank { throw LlmException("Claude CLI 응답이 비어있습니다.") }
    }

    /** 블로킹 모드: 전체 응답을 한 번에 읽음 */
    private fun readBlocking(process: Process): String {
        val sb = StringBuilder()

        val readerThread = Thread {
            try {
                sb.append(process.inputStream.bufferedReader().readText())
            } catch (_: Exception) { }
        }
        readerThread.isDaemon = true
        readerThread.start()

        val finished = process.waitFor(TIMEOUT_SECONDS, TimeUnit.SECONDS)
        readerThread.join(2_000)

        if (!finished) {
            process.destroyForcibly()
            throw LlmException(
                "Claude CLI 응답 시간 초과 (${TIMEOUT_SECONDS}초)\n" +
                "› 네트워크 상태나 claude 로그인 여부를 확인해주세요."
            )
        }

        val output = sb.toString().trim()
        val exitCode = runCatching { process.exitValue() }.getOrDefault(-1)
        if (exitCode != 0) {
            throw LlmException("Claude CLI 오류 (exit $exitCode):\n${output.take(300)}")
        }

        return output.ifBlank { throw LlmException("Claude CLI 응답이 비어있습니다.") }
    }

    private fun checkNotLoggedIn(output: String) {
        if (output.contains("Not logged in", ignoreCase = true) ||
            output.contains("Please run /login", ignoreCase = true) ||
            output.contains("not authenticated", ignoreCase = true)
        ) {
            throw LlmException(
                "Claude CLI 로그인이 필요합니다.\n" +
                "› Settings → Tools → Niku Agent에서 '터미널에서 로그인' 버튼을 클릭하거나\n" +
                "› 터미널에서 `claude /login`을 실행해주세요."
            )
        }
    }

    companion object {
        private const val TIMEOUT_SECONDS = 120L

        private val CANDIDATE_PATHS = listOf(
            "/usr/local/bin/claude",
            "/opt/homebrew/bin/claude",
            "${System.getProperty("user.home")}/.local/bin/claude",
            "${System.getProperty("user.home")}/.npm/bin/claude",
            "${System.getProperty("user.home")}/.nvm/bin/claude",
        )

        fun findBinary(): String? {
            for (path in CANDIDATE_PATHS) {
                if (File(path).canExecute()) return path
            }
            return try {
                val proc = ProcessBuilder("which", "claude")
                    .redirectErrorStream(true)
                    .start()
                proc.waitFor(5, TimeUnit.SECONDS)
                val result = proc.inputStream.bufferedReader().readText().trim()
                result.takeIf { it.isNotBlank() && File(it).canExecute() }
            } catch (_: Exception) {
                null
            }
        }

        fun checkVersion(binaryPath: String): String? {
            return try {
                val proc = ProcessBuilder(binaryPath, "--version")
                    .redirectErrorStream(true)
                    .start()
                proc.waitFor(5, TimeUnit.SECONDS)
                proc.inputStream.bufferedReader().readText().trim().takeIf { it.isNotBlank() }
            } catch (_: Exception) {
                null
            }
        }
    }
}

File: vscode-extension/src/cliLlmClient.ts
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

const TIMEOUT_MS = 120_000;

const CANDIDATE_PATHS = [
  '/usr/local/bin/claude',
  '/opt/homebrew/bin/claude',
  path.join(os.homedir(), '.local/bin/claude'),
  path.join(os.homedir(), 'Library/Application Support/Claude/claude-code-vm/2.1.78/claude'),
  path.join(os.homedir(), 'Library/Application Support/Claude/claude-code/2.1.78/claude'),
  path.join(os.homedir(), 'Library/Application Support/Claude/claude-code-vm/2.1.72/claude'),
  path.join(os.homedir(), 'Library/Application Support/Claude/claude-code/2.1.72/claude'),
  path.join(os.homedir(), '.npm/bin/claude'),
  path.join(os.homedir(), '.nvm/bin/claude'),
];

export class LlmException extends Error {
  constructor(message: string, public readonly cause?: Error) {
    super(message);
    this.name = 'LlmException';
  }
}

export class CliLlmClient {
  constructor(
    private readonly binaryPath: string,
    private readonly model?: string,
  ) {}

  /**
   * Claude CLI 를 subprocess 로 실행하고 응답을 반환한다.
   * onChunk 가 제공되면 실시간 스트리밍 콜백을 호출한다.
   */
  complete(prompt: string, onChunk?: (accumulated: string) => void): Promise<string> {
    return new Promise((resolve, reject) => {
      const args = ['--print'];
      if (this.model) { args.push('--model', this.model); }

      let proc: ReturnType<typeof spawn>;
      try {
        proc = spawn(this.binaryPath, args, { stdio: ['pipe', 'pipe', 'pipe'] });
      } catch (err: unknown) {
        return reject(new LlmException(
          `Claude CLI 실행 실패: ${(err as Error).message}\n` +
          '› claude CLI가 설치되어 있는지 확인해주세요.',
          err as Error,
        ));
      }

      // stdin 으로 프롬프트 전달 후 EOF
      proc.stdin!.write(prompt, 'utf8');
      proc.stdin!.end();

      const chunks: string[] = [];
      let accumulated = '';

      proc.stdout!.on('data', (data: Buffer) => {
        const text = data.toString('utf8');
        chunks.push(text);
        accumulated += text;
        onChunk?.(accumulated);
      });

      const timer = setTimeout(() => {
        proc.kill('SIGKILL');
        if (accumulated.trim()) {
          onChunk?.(`${accumulated}\n\n⚠️ 응답 시간 초과로 부분 결과만 표시됩니다.`);
          resolve(accumulated.trim());
        } else {
          reject(new LlmException(`Claude CLI 응답 시간 초과 (${TIMEOUT_MS / 1000}초)\n› 네트워크 상태나 claude 로그인 여부를 확인해주세요.`));
        }
      }, TIMEOUT_MS);

      proc.on('close', (code) => {
        clearTimeout(timer);
        const output = accumulated.trim();

        try {
          checkNotLoggedIn(output);
        } catch (e) {
          return reject(e);
        }

        if (code !== 0 && !output) {
          return reject(new LlmException(`Claude CLI 오류 (exit ${code}). Settings에서 버전 확인 버튼으로 상태를 확인해주세요.`));
        }

        if (!output) {
          return reject(new LlmException('Claude CLI 응답이 비어있습니다.'));
        }

        resolve(output);
      });

      proc.on('error', (err) => {
        clearTimeout(timer);
        reject(new LlmException(`Claude CLI 실행 실패: ${err.message}`, err));
      });
    });
  }

  // ── Static helpers ──────────────────────────────────────────────

  static findBinary(): string | null {
    for (const p of CANDIDATE_PATHS) {
      try {
        fs.accessSync(p, fs.constants.X_OK);
        return p;
      } catch { /* not found */ }
    }
    // which claude
    try {
      const { execSync } = require('child_process') as typeof import('child_process');
      const result = execSync('which claude', { encoding: 'utf8', timeout: 5000 }).trim();
      if (result) {
        try { fs.accessSync(result, fs.constants.X_OK); return result; } catch { /* skip */ }
      }
    } catch { /* skip */ }
    return null;
  }

  static checkVersion(binaryPath: string): string | null {
    try {
      const { execSync } = require('child_process') as typeof import('child_process');
      return execSync(`"${binaryPath}" --version`, { encoding: 'utf8', timeout: 5000 }).trim() || null;
    } catch {
      return null;
    }
  }

  static checkLoginStatus(): string {
    try {
      const claudeJson = path.join(os.homedir(), '.claude.json');
      if (!fs.existsSync(claudeJson)) {
        return '❌ 로그인되어 있지 않습니다.';
      }
      const content = fs.readFileSync(claudeJson, 'utf8');
      const ok = content.includes('"oauthAccount"') &&
                 content.includes('"accountUuid"') &&
                 content.includes('"emailAddress"');
      return ok ? '✅ 로그인 상태 정상' : '❌ 로그인되어 있지 않습니다.';
    } catch (e: unknown) {
      return `❌ 확인 실패: ${(e as Error).message}`;
    }
  }
}

function checkNotLoggedIn(output: string): void {
  const outputSignal =
    /not logged in/i.test(output) ||
    /please run \/login/i.test(output) ||
    /not authenticated/i.test(output);

  let fileSignal = false;
  try {
    const claudeJson = path.join(os.homedir(), '.claude.json');
    if (!fs.existsSync(claudeJson)) {
      fileSignal = true;
    } else {
      const c = fs.readFileSync(claudeJson, 'utf8');
      fileSignal = !c.includes('"oauthAccount"') || !c.includes('"accountUuid"');
    }
  } catch { /* ignore */ }

  if (outputSignal || fileSignal) {
    throw new LlmException(
      '로그인이 필요합니다.\n' +
      '› 터미널에서 claude 를 실행하고 /login 명령을 입력해주세요.',
    );
  }
}
